memory/src/main/scala/filodb.memory/format/vectors/Histogram.scala

```diff
@@ -0,0 +1,295 @@
+  val BucketTypeExplicit  = 0x03.toByte
+
+  val OffsetBucketType = 2
+  val OffsetNumBuckets = 0
+  val OffsetBucketDetails = 3
+
+  /**
+   * Creates the right HistogramBuckets from a binary definition.  NOTE: length field not included here
+   * The first two bytes of any binary bucket schema definition are the number of buckets.
+   * The third byte is the bucket scheme ID.
+   */
+  def apply(binaryBucketsDef: Array[Byte]): HistogramBuckets =
+    apply(binaryBucketsDef, UnsafeUtils.arayOffset, binaryBucketsDef.size)
+
+  def apply(bucketsDef: BinaryRegion.NativePointer, numBytes: Int): HistogramBuckets =
+    apply(UnsafeUtils.ZeroPointer.asInstanceOf[Array[Byte]], bucketsDef, numBytes)
+
+  def apply(bucketsDefBase: Array[Byte], bucketsDefOffset: Long, numBytes: Int): HistogramBuckets = {
+    UnsafeUtils.getByte(bucketsDefBase, bucketsDefOffset + OffsetBucketType) match {
+      case BucketTypeGeometric   =>
+        GeometricBuckets(UnsafeUtils.getDouble(bucketsDefBase, bucketsDefOffset + OffsetBucketDetails),
+                         UnsafeUtils.getDouble(bucketsDefBase, bucketsDefOffset + OffsetBucketDetails + 8),
+                         UnsafeUtils.getShort(bucketsDefBase, bucketsDefOffset + OffsetNumBuckets).toInt)
+      case BucketTypeGeometric_1 =>
+        GeometricBuckets_1(UnsafeUtils.getDouble(bucketsDefBase, bucketsDefOffset + OffsetBucketDetails),
+                           UnsafeUtils.getDouble(bucketsDefBase, bucketsDefOffset + OffsetBucketDetails + 8),
+                           UnsafeUtils.getShort(bucketsDefBase, bucketsDefOffset + OffsetNumBuckets).toInt)
+      case _ =>
+        // Probably due to empty histogram vector, just return some default
+        binaryBuckets64
+    }
+  }
+
+  /**
+   * Creates a binary bucket definition for a geometric series of histogram buckets.
+   * That means each bucket contains values <= (firstBucket) * (multipler) ^ (bucketNo)
+   * where bucketNo starts at 0 and goes till (numBuckets - 1)
+   * @param firstBucket initial bucket value
+   * @param multiplier the geometric multiplier between buckets
+   * @param numBuckets the total number of buckets
+   * @param minusOne if true, 1 is subtracted from each geometric bucket from the formula above
+   */
+  final def geometricBucketDef(firstBucket: Double,
+                               multiplier: Double,
+                               numBuckets: Int,
+                               minusOne: Boolean = false): Array[Byte] = {
+    require(numBuckets < 65536, s"Too many buckets: $numBuckets")
+    val bytes = new Array[Byte](19)
+    bytes(OffsetBucketType) = if (minusOne) BucketTypeGeometric_1 else BucketTypeGeometric
+    UnsafeUtils.setShort(bytes, UnsafeUtils.arayOffset + OffsetNumBuckets, numBuckets.toShort)
+    UnsafeUtils.setDouble(bytes, UnsafeUtils.arayOffset + OffsetBucketDetails, firstBucket)
+    UnsafeUtils.setDouble(bytes, UnsafeUtils.arayOffset + OffsetBucketDetails + 8, multiplier)
```

Comment:
We can def do this when we know the input histogram type.  Prom's API actually has geometric vs other bucket schemes.  It would be super if we could somehow get Prom to expose that in the metric endpoint, otherwise we have to guess which is not going to be fast....